hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/predicate/DatePredicateBuilder.java

```diff
@@ -282,4 +139,226 @@ public enum ColumnEnum {
+			if (theOperation == SearchFilterParser.CompareOperation.lt
+					|| theOperation == SearchFilterParser.CompareOperation.le) {
+				// use lower bound first
+				handleLessThanAndLessThanOrEqualTo();
+			} else if (theOperation == SearchFilterParser.CompareOperation.gt
+					|| theOperation == SearchFilterParser.CompareOperation.ge) {
+				// use upper bound first, e.g value between 6 and 10
+				handleGreaterThanAndGreaterThanOrEqualTo();
+			} else if (theOperation == SearchFilterParser.CompareOperation.ne) {
+				if ((lowerBoundInstant == null) || (upperBoundInstant == null)) {
+					throw new InvalidRequestException(Msg.code(1254)
+							+ "lowerBound and/or upperBound value not correctly specified for compare theOperation");
+				}
+				lessThan = DatePredicateBuilder.this.createPredicate(
+						lowValueField, ParamPrefixEnum.LESSTHAN, genericLowerBound);
+				greaterThan = DatePredicateBuilder.this.createPredicate(
+						highValueField, ParamPrefixEnum.GREATERTHAN, genericUpperBound);
+				lowerBoundCondition = ComboCondition.or(lessThan, greaterThan);
+			} else if ((theOperation == SearchFilterParser.CompareOperation.eq)
+					|| (theOperation == SearchFilterParser.CompareOperation.sa)
+					|| (theOperation == SearchFilterParser.CompareOperation.eb)
+					|| (theOperation == null)) {
+
+				handleEqualToCompareOperator();
+			} else {
+				throw new InvalidRequestException(Msg.code(1255)
+						+ String.format("Unsupported operator specified, operator=%s", theOperation.name()));
+			}
+
+			if (isOrdinalComparison()) {
+				ourLog.trace("Ordinal date range is {} - {} ", lowerBoundAsOrdinal, upperBoundAsOrdinal);
+			} else {
+				ourLog.trace("Date range is {} - {}", lowerBoundInstant, upperBoundInstant);
+			}
+
+			if (lowerBoundCondition != null && upperBoundCondition != null) {
+				return (ComboCondition.and(lowerBoundCondition, upperBoundCondition));
+			} else if (lowerBoundCondition != null) {
+				return (lowerBoundCondition);
+			} else {
+				return (upperBoundCondition);
+			}
+		}
+
+		private void handleEqualToCompareOperator() {
+			if (lowerBoundInstant != null && upperBoundInstant != null) {
+				// both upper and lower bound
+				// lowerbound; :lowerbound <= low_field <= :upperbound
+				greaterThan = ComboCondition.and(
+						DatePredicateBuilder.this.createPredicate(
+								lowValueField, ParamPrefixEnum.GREATERTHAN_OR_EQUALS, genericLowerBound),
+						DatePredicateBuilder.this.createPredicate(
+								lowValueField, ParamPrefixEnum.LESSTHAN_OR_EQUALS, genericUpperBound));
+				// upperbound; :lowerbound <= high_field <= :upperbound
+				lessThan = ComboCondition.and(
+						DatePredicateBuilder.this.createPredicate(
+								highValueField, ParamPrefixEnum.LESSTHAN_OR_EQUALS, genericUpperBound),
+						DatePredicateBuilder.this.createPredicate(
+								highValueField, ParamPrefixEnum.GREATERTHAN_OR_EQUALS, genericLowerBound));
```

Comment:
I don't know why there's greaterThan, lessThan as well as lowerBoundCondition and upperBoundCondition.  But this might make more sense to me rearranged to make it clear we apply the same condition to both columns:
```suggestion
				greaterThan = ComboCondition.and(
						DatePredicateBuilder.this.createPredicate(
								lowValueField, ParamPrefixEnum.GREATERTHAN_OR_EQUALS, genericLowerBound),
						DatePredicateBuilder.this.createPredicate(
								highValueField, ParamPrefixEnum.GREATERTHAN_OR_EQUALS, genericLowerBound));
				// upperbound; :lowerbound <= high_field <= :upperbound
				lessThan = ComboCondition.and(
						DatePredicateBuilder.this.createPredicate(
								lowValueField, ParamPrefixEnum.LESSTHAN_OR_EQUALS, genericUpperBound),
						DatePredicateBuilder.this.createPredicate(
								highValueField, ParamPrefixEnum.LESSTHAN_OR_EQUALS, genericUpperBound));
```